### ISSUE TITLE
Unexpected crash

### ISSUE THREAD

**Issue by quyleanh** [krlvm/PowerTunnel-Android#99]
App version: 2.0 (100)
Core version: 2.0 (100)
OS Version: 22 (API 30)

```
android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{908511 u0 io.github.krlvm.powertunnel.android/.services.TunnelingVpnService}
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2005)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by quyleanh** [comment]
I think the problem is I enable both Firewall and Adblock plugins.

**Comment by krlvm** [comment]
Can you provide more complete log? You can enable log viewer in settings.

**Comment by quyleanh** [comment]
Here is the log.

```
--------- beginning of crash
12-28 20:29:56.144 11084 11084 E AndroidRuntime: FATAL EXCEPTION: main
12-28 20:29:56.144 11084 11084 E AndroidRuntime: Process: io.github.krlvm.powertunnel.android, PID: 11084
12-28 20:29:56.144 11084 11084 E AndroidRuntime: android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{908511 u0 io.github.krlvm.powertunnel.android/.services.TunnelingVpnService}
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2005)
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7660)
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
12-28 20:29:56.144 11084 11084 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: FATAL EXCEPTION: main
12-28 20:33:30.061 18952 18952 E AndroidRuntime: Process: io.github.krlvm.powertunnel.android, PID: 18952
12-28 20:33:30.061 18952 18952 E AndroidRuntime: android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{77eed75 u0 io.github.krlvm.powertunnel.android/.services.TunnelingVpnService}
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2005)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7660)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
12-28 20:33:30.061 18952 18952 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
--------- beginning of main
12-28 21:41:15.630 19222 26152 I i*.g*.k*.p*.PowerTunnel: Registered plugin 'Firewall' [firewall] v1.0 (1) by krlvm
12-28 21:41:15.640 19222 26152 I i*.g*.k*.p*.PowerTunnel: Registered plugin 'AdBlock' [adblock] v1.0-vn (1) by krlvm
12-28 21:41:15.645 19222 26152 I i*.g*.k*.p*.PowerTunnel: Registered plugin 'LibertyTunnel' [libertytunnel] v1.0 (1) by krlvm
12-28 21:41:15.639 19222 19222 W io.github.krlvm.powertunnel.android: type=1400 audit(0.0:45218): avc: granted { execute } for comm=50726F787920426F6F747374726170 path="/data/data/io.github.krlvm.powertunnel.android/files/plugins/oat/arm64/DNSResolver-Android.odex" dev="sda21" ino=5637461 scontext=u:r:untrusted_app:s0:c187,c257,c512,c768 tcontext=u:object_r:app_data_file:s0:c187,c257,c512,c768 tclass=file app=io.github.krlvm.powertunnel.android
12-28 21:41:15.657 19222 26152 I i*.g*.k*.p*.PowerTunnel: Registered plugin 'DNS' [dns] v1.0 (1) by krlvm
12-28 21:41:15.660 19222 26152 W FirewallPlugin: Filtration list is empty
12-28 21:41:15.661 19222 26152 I FirewallPlugin: Filtration enabled for 0 hosts (0 hosts are excluded) in BLACKLIST mode with DNS filtration type
12-28 21:41:15.663 19222 26152 I AdBlock : Loading filters from mirror...
12-28 21:41:25.607 19222 19222 I TunnelingVpnService: Stopping...
12-28 21:41:25.616 19222 19222 E AndroidRuntime: FATAL EXCEPTION: main
12-28 21:41:25.616 19222 19222 E AndroidRuntime: Process: io.github.krlvm.powertunnel.android, PID: 19222
12-28 21:41:25.616 19222 19222 E AndroidRuntime: android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{9dcd58c u0 io.github.krlvm.powertunnel.android/.services.TunnelingVpnService}
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2005)
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7660)
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
12-28 21:41:25.616 19222 19222 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
12-28 21:41:25.656 19222 19222 I Process : Sending signal. PID: 19222 SIG: 9
12-28 21:41:27.647 26162 26162 E rtunnel.androi: Not starting debugger since process cannot load the jdwp agent.
12-28 21:41:27.679 26162 26162 I MultiDex: VM with version 2.1.0 has multidex support
12-28 21:41:27.679 26162 26162 I MultiDex: Installing application
12-28 21:41:27.679 26162 26162 I MultiDex: VM has multidex support, MultiDex support library is disabled.
12-28 21:41:27.685 26162 26162 W rtunnel.androi: Accessing hidden field Ljava/util/Collections$SynchronizedCollection;->mutex:Ljava/lang/Object; (greylist-max-o, reflection, denied)
12-28 21:41:27.686 26162 26162 W rtunnel.androi: Accessing hidden method Ljava/util/Collections$SynchronizedSet;-><init>(Ljava/util/Set;Ljava/lang/Object;)V (greylist-max-o, reflection, denied)
12-28 21:41:27.686 26162 26162 W rtunnel.androi: Accessing hidden method Ljava/util/Collections$SynchronizedCollection;-><init>(Ljava/util/Collection;Ljava/lang/Object;)V (greylist-max-o, reflection, denied)
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: QUALCOMM build                   : 95db91f, Ifbc588260a
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Build Date                       : 09/24/20
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: OpenGL ES Shader Compiler Version: EV031.32.02.01
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Local Branch                     : mybrancheafe5b6d-fb5b-f1b0-b904-5cb90179c3e0
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Remote Branch                    : quic/gfx-adreno.lnx.1.0.r114-rel
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Remote Branch                    : NONE
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Reconstruct Branch               : NOTHING
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Build Config                     : S P 10.0.7 AArch64
12-28 21:41:27.719 26162 26179 I AdrenoGLES-0: Driver Path                      : /vendor/lib64/egl/libGLESv2_adreno.so
12-28 21:41:27.725 26162 26179 I AdrenoGLES-0: PFP: 0x016ee187, ME: 0x00000000
12-28 21:41:27.736 26162 26179 I Gralloc4: mapper 4.x is not supported
12-28 21:41:27.736 26162 26179 W Gralloc3: mapper 3.x is not supported
12-28 21:41:33.113 26162 26162 W RecyclerView: No adapter attached; skipping layout
12-28 21:41:43.749 26162 26162 W Choreographer: Frame time is 0.309875 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
```

**Comment by krlvm** [comment]
I see, it no longer crashes?

**Comment by quyleanh** [comment]
If I turn of Firewall plugin, then yes. It no longer crashes.

**Comment by krlvm** [comment]
Can't reproduce on two devices, are you sure you've installed the Android version of Firewall plugin?

**Comment by quyleanh** [comment]
Yes, I installed correct version of Firewall (with Android version).
I can't reproduce now. I think it might conflict with Wireguard somehow.

I will close this issue. Thank you for your time.

**Comment by krlvm** [comment]
No, it can't conflict with Wireguard, the problem is that .jar file should be made compatible with Android, so I always attach two jar files for Android and PC.

**Comment by quyleanh** [comment]
![Screenshot_20211228-225023_PowerTunnel](https://user-images.githubusercontent.com/9365035/147584413-16325462-4676-4ee1-86e3-9573fae54041.png)

I use correct version, right? The bug happens again. I will try to reproduce and report later.

**Comment by krlvm** [comment]
Yes, that version is correct

**Comment by krlvm** [comment]
Should be fixed in [version 2.0.1](https://github.com/krlvm/PowerTunnel-Android/releases/v2.0.1), though I believe it comes from a plugin crash.
